Fix crash caused by invalid `mix_rate` assignment due to bogus project settings.

Eoin-ONeill-Yokai opened the pull request:
I noticed this bug come in that was related to code I was already looking over for other reasons, so I decided to fix it.

We'll just default to a sensible value in the case that a user has somehow managed to modify the configuration file incorrectly. 44100hz feels like the appropriate default, but if there's a constant somewhere defined let me know and I'll change it to that. Otherwise, I'll change the code to use a constant so the fallback makes sense on a system-by-system basis and is easy to change.

It seems like, to me, there shouldn't be a case where we're loading the audio driver at all with an improper `mix_rate`, but I could be wrong. In which case, we will simply make exceptions for the `divide by zero` error. 

Closes #69819

Eoin-ONeill-Yokai committed:
Fix crash caused by invalid mix_rate assignment due to bogus
project settings.

We'll default to a sensible value in the case that a user has
somehow managed to modify the configuration file incorrectly.

Closes 69819

adamscott commented on drivers/pulseaudio/audio_driver_pulseaudio.cpp:
```diff
@@ -290,6 +290,11 @@ Error AudioDriverPulseAudio::init() {
 
+	// In the case of invalid mix rate, let's default to a sensible value..
+	if (mix_rate <= 0) {
+		mix_rate = 44100;
```
I suggest to add a warning that tells the user that their `audio/driver/mix_rate` setting is invalid and was set to 44 100 KHz.

adamscott commented on drivers/xaudio2/audio_driver_xaudio2.cpp:
```diff
@@ -44,6 +44,12 @@ Error AudioDriverXAudio2::init() {
+
+	// In the case of invalid mix rate, let's default to a sensible value..
+	if (mix_rate <= 0) {
+		mix_rate = 44100;
```
Same comment about the warning here.

akien-mga commented:
I think there's a `DEFAULT_MIX_RATE` constant somewhere in AudioServer or AudioDriver which you could use.

I haven't checked but there are more audio driver implementations which might need the same treatment (alsa, coreaudio, wasapi, dummy).

Eoin-ONeill-Yokai commented:
> I think there's a `DEFAULT_MIX_RATE` constant somewhere in AudioServer or AudioDriver which you could use.

Yes reusing that is a good plan. I've moved the declaration to a public section of the class.

> I haven't checked but there are more audio driver implementations which might need the same treatment (alsa, coreaudio, wasapi, dummy).

Yeah there was more to do there. I have added all the extra logic to the other platforms where appropriate. I skipped the warning in the dummy class, as it likely isn't as important there since users won't likely run into that specific driver all that often and it seems like it's for advance users anyway.

akien-mga commented on drivers/alsa/audio_driver_alsa.cpp:
```diff
@@ -44,7 +44,9 @@ extern int initialize_pulse(int verbose);
 Error AudioDriverALSA::init_output_device() {
-	mix_rate = GLOBAL_GET("audio/driver/mix_rate");
+	StringName audio_driver_setting = "audio/driver/mix_rate";
+	mix_rate = GLOBAL_GET(audio_driver_setting);
```
Shouldn't this one get a validation check too?

akien-mga reviewed (commented):
Since it's the same logic everywhere, it would make sense IMO to move this to a function like `AudioDriverManager::get_default_mix_rate()` so you don't need to duplicate the logic in each driver.

Eoin-ONeill-Yokai commented:
> so you don't need to duplicate the logic in each driver.

I was wondering about this. Would it be ok to make a function that reads from the configuration and returns the default if it's invalid? It seems pretty specific and would only be useful in the constructor, but for a protected method perhaps it would be ok as long as people don't confuse it with `get_mix_rate` or similar functions. Like you said, it would cut down on repetition.

akien-mga commented:
Yeah that's ok as a helper function. You can make it protected and even start with an underscore to make it clear that it's not part of the public API.

akien-mga commented:
Thanks!

YuriSizov commented:
Cherry-picked for 4.0.3.
